scripts/gb_model/process_fes_hydrogen_data.py

```diff
@@ -0,0 +1,576 @@
+        2. Group mapped data by bus (region) and year
+        3. Calculate regional distribution proportions for each year
+        4. Identify and aggregate unmapped capacities (NaN bus values)
+        5. Redistribute unmapped capacities based on regional distribution patterns
+        6. Combine mapped and redistributed capacities into final dataset
+    """
+    # Read regional GB data
+    regional_gb_data = pd.read_csv(regional_gb_data_path)
+
+    # Select regional hydrogen electrolysis data
+    electrolysis_data = regional_gb_data[
+        (regional_gb_data["Technology Detail"].str.lower() == "hydrogen electrolysis")
+    ]
+
+    # Calculate regional grid-connected electrolysis capacities
+    regional_grid_electrolysis_capacities = electrolysis_data.groupby(["bus", "year"])[
+        "data"
+    ].sum()
+
+    # Calculate regional distribution of grid-connected electrolysis capacities
+    electrolysis_distribution = get_regional_distribution(
+        regional_grid_electrolysis_capacities
+    )
+
+    # Map unmapped grid-connected electrolysis data
+    unmapped_grid_electrolysis_capacities = (
+        electrolysis_data[electrolysis_data["bus"].isnull()]
+        .groupby("year")["data"]
+        .sum()
+    )
+    unmapped_grid_electrolysis_capacities = (
+        unmapped_grid_electrolysis_capacities * electrolysis_distribution
+    )
+
+    # Combine mapped and unmapped grid-connected electrolysis data
+    grid_electrolysis_capacities = (
+        regional_grid_electrolysis_capacities + unmapped_grid_electrolysis_capacities
+    )
+
+    # Rename series to 'p_nom'
+    grid_electrolysis_capacities.name = "p_nom"
+
+    return grid_electrolysis_capacities
+
+
+def parse_supply_data(
+    supply_sheets: list,
+    supply_sheets_mapping: dict,
+    year_range: list,
+) -> pd.DataFrame:
+    """
+    Parse the hydrogen supply data to the required format.
+
+    Args:
+        supply_sheets (list): List of file paths to supply sheet CSV files
+        supply_sheets_mapping (dict): Nested mapping of fes_year -> data type -> sheet_name
+        year_range (list): Two-element list [start_year, end_year] defining the year range to include
+
+    Returns:
+        pd.DataFrame: Processed hydrogen supply data with MultiIndex ['bus', 'year'] and 'data' values in MWh.
+    """
+    # Retrieve annual supply data from sheets
+    for year, sheets in supply_sheets_mapping.items():
+        for data_type, sheet_name in sheets.items():
+            # Find the corresponding file in the supply_sheets
+            if data_type == "hydrogen_supply":
+                sheet_file = next(f for f in supply_sheets if sheet_name in f)
+
+    # Read the supply sheet data
+    supply_data = pd.read_csv(sheet_file)
+    supply_data = supply_data.apply(_strip_str)
+
+    # Select exogeneous hydrogen supply data
+    supply_data = supply_data[
+        supply_data["carrier"].str.lower().isin(exogeneous_supply_list)
```

Comment:
how does this know what `exogeneous_supply_list` is if it hasn't received it as an input to the method?